test/namedtuple.jl

```diff
@@ -0,0 +1,161 @@
+
+@test keys((a=1, b=2, c=3)) == (:a, :b, :c)
+@test keys(NamedTuple()) == ()
+@test keys((a=1,)) == (:a,)
+@test values((a=1, b=2, c=3)) == (1, 2, 3)
+@test values(NamedTuple()) == ()
+@test values((a=1,)) == (1,)
+@test haskey((a=1, b=2, c=3), :a)
+@test !haskey(NamedTuple(), :a)
+@test !haskey((a=1,), :b)
+@test get((a=1, b=2, c=3), :a, 0) == 1
+@test get(NamedTuple(), :a, 0) == 0
+@test get((a=1,), :b, 0) == 0
+@test get(()->0, (a=1, b=2, c=3), :a) == 1
+@test get(()->0, NamedTuple(), :a) == 0
+@test get(()->0, (a=1,), :b) == 0
+
+# syntax errors
+
+@test Meta.lower(Main, parse("(a=1, 0)")) == Expr(:error, "invalid named tuple element \"0\"")
+@test Meta.lower(Main, parse("(a=1, f(x))")) == Expr(:error, "invalid named tuple element \"f(x)\"")
+@test Meta.lower(Main, parse("(a=1,a=2)")) == Expr(:error, "field name \"a\" repeated in named tuple")
+@test Meta.lower(Main, parse("(a=1,b=0,a=2)")) == Expr(:error, "field name \"a\" repeated in named tuple")
+@test Meta.lower(Main, parse("(c=1,a=1,b=0,a=2)")) == Expr(:error, "field name \"a\" repeated in named tuple")
+
+@test parse("(;)") == quote end
+@test Meta.lower(Main, parse("(1,;2)")) == Expr(:error, "unexpected semicolon in tuple")
+
+# splatting and implicit naming syntax
+let d = [:a=>1, :b=>2, :c=>3]   # use an array to preserve order
+    @test (d..., a=10) == (a=10, b=2, c=3)
+    @test (a=0, b=0, z=1, d..., x=4, y=5) == (a=1, b=2, z=1, c=3, x=4, y=5)
+#=
+    x = 10
+    t = (x=1, y=20)
+    @test (;d...) == (a=1, b=2, c=3)
+    @test (;d..., x, t.y) == (a=1, b=2, c=3, x=10, y=20)
+    @test (;d..., :z=>20) == (a=1, b=2, c=3, z=20)
+    @test (;a=10, d..., :c=>30) == (a=1, b=2, c=30)
+    @test (;t.x, t.y) == (x=1, y=20)
+    @test (b=1, t.y) == (b=1, y=20)
+    y = (w=30, z=40)
+    @test (;t..., y...) == (x=1, y=20, w=30, z=40)
+    @test (;t..., y=0, y...) == (x=1, y=0, w=30, z=40)
+=#
+end
+#=
+@test Meta.lower(Main, parse("(; f(x))")) == Expr(:error, "invalid named tuple element \"f(x)\"")
+@test Meta.lower(Main, parse("(;1=0)")) == Expr(:error, "invalid named tuple field name \"1\"")
+@test Meta.lower(Main, parse("(c=1,a=1,b=0,d.a)")) == Expr(:error, "field name \"a\" repeated in named tuple")
+@test Meta.lower(Main, parse("(c=1,a=1,b=0,c)")) == Expr(:error, "field name \"c\" repeated in named tuple")
+@test Meta.lower(Main, parse("(;d.c,c)")) == Expr(:error, "field name \"c\" repeated in named tuple")
+=#
+
+# inference tests
+
+namedtuple_get_a(x) = x.a
+@test Base.return_types(namedtuple_get_a, (NamedTuple,)) == Any[Any]
+@test Base.return_types(namedtuple_get_a, (typeof((b=1,a="")),)) == Any[String]
+
+namedtuple_fieldtype_a(x) = fieldtype(typeof(x), :a)
+@test Base.return_types(namedtuple_fieldtype_a, (NamedTuple,)) == Any[Type]
+@test Base.return_types(namedtuple_fieldtype_a, (typeof((b=1,a="")),)) == Any[Type{String}]
+
+namedtuple_nfields(x) = nfields(x) === 0 ? 1 : ""
+@test Union{Int,String} <: Base.return_types(namedtuple_nfields, (NamedTuple,))[1]
```

Comment:
can we do an exact test?

Reply:
I like checking soundness, since then we don't get random tests breaking from e.g. changing the heuristics in `tmerge`.